File: src/server2/src/util/result.ts
interface BaseResult<T, E> {
    isOk: boolean
    isErr: boolean
    value?: T
    error?: E
}

export class Ok<T, E> implements BaseResult<T, E> {
    public readonly isOk: boolean
    public readonly isErr: boolean
    public readonly error?: E | undefined
    public readonly value: T | undefined

    constructor(v?: T) {
        this.isOk = true
        this.isErr = false
        this.error = undefined
        this.value = v
    }
}

export class Err<T, E> implements BaseResult<T, E> {
    public readonly isOk: boolean
    public readonly isErr: boolean
    public readonly error?: E
    public readonly value: T | undefined

    constructor(error: E) {
        this.isOk = false
        this.isErr = true
        this.error = error
        this.value = undefined
    }
}

type Result<T, E> = Ok<T, E> | Err<T, E>

export default Result

File: src/cserver/LunchPicker.API/Common/Result.cs
namespace LunchPicker.API
{
    public class Result<TOk, TErr>
    {
        public Result(TOk ok)
        {
            Value = ok;
        }

        public Result(TErr err)
        {
            Error = err;
        }

        public TOk Value { get; private set; }

        public TErr Error { get; private set; }

        public bool IsOk => Value != null;

        public bool IsError => Error != null;
    }
}